File: src/main/java/com/toy/shopwebmvc/service/ItemService.java
package com.toy.shopwebmvc.service;

import com.toy.shopwebmvc.dto.request.ItemSaveRequest;
import com.toy.shopwebmvc.dto.response.ItemResponse;
import com.toy.shopwebmvc.repository.ItemRepository;
import com.toy.shopwebmvc.domain.Item;
import com.toy.shopwebmvc.exception.CommonException;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import static com.toy.shopwebmvc.constant.ApiResponseCode.*;

@Service
@Transactional(readOnly = true)
@RequiredArgsConstructor
public class ItemService {

    private final ItemRepository itemRepository;

    @Transactional
    public ItemResponse saveItem(ItemSaveRequest itemSaveRequest) {
        Item item = Item.create(
                itemSaveRequest.name(),
                itemSaveRequest.description(),
                itemSaveRequest.price(),
                itemSaveRequest.quantity(),
                itemSaveRequest.category()
        );

        itemRepository.save(item);

        return new ItemResponse(item);
    }

    public Page<ItemResponse> findItems(Pageable pageable, String name) {
        return itemRepository.findAllWithQuerydsl(pageable, name)
                .map(ItemResponse::new);
    }

    public ItemResponse findItem(Long id) {
        return itemRepository.findById(id)
                .map(ItemResponse::new)
                .orElseThrow(() -> new CommonException(DATA_NOT_FOUND));
    }

    @Transactional
    public ItemResponse updateItem(Long id, ItemSaveRequest itemSaveRequest) {
        Item item = itemRepository.findById(id)
                .map(findItem -> {
                    findItem.update(
                            itemSaveRequest.name(),
                            itemSaveRequest.description(),
                            itemSaveRequest.price(),
                            itemSaveRequest.quantity(),
                            itemSaveRequest.category()
                    );

                    return findItem;
                })
                .orElseGet(() -> Item.create(
                        itemSaveRequest.name(),
                        itemSaveRequest.description(),
                        itemSaveRequest.price(),
                        itemSaveRequest.quantity(),
                        itemSaveRequest.category()
                ));

        return new ItemResponse(item);
    }

    @Transactional
    public void deleteItem(Long id) {
        itemRepository.deleteById(id);
    }
}

File: spring-web-kotlin/src/main/kotlin/spring/web/kotlin/domain/item/service/ItemService.kt
package spring.web.kotlin.domain.item.service

import org.springframework.data.domain.Pageable
import org.springframework.data.repository.findByIdOrNull
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import spring.web.kotlin.domain.item.Item
import spring.web.kotlin.domain.item.dto.ItemResponse
import spring.web.kotlin.domain.item.dto.ItemSaveRequest
import spring.web.kotlin.domain.item.repository.ItemRepository
import spring.web.kotlin.global.common.ResponseMessage
import spring.web.kotlin.global.exception.DomainException

@Service
@Transactional(readOnly = true)
class ItemService(
    private val itemRepository: ItemRepository,
) {
    @Transactional
    fun saveItem(itemSaveRequest: ItemSaveRequest): ItemResponse {
        val item = itemRepository.save(
            Item.create(
                itemSaveRequest.name,
                itemSaveRequest.description,
                itemSaveRequest.price,
                itemSaveRequest.quantity,
                itemSaveRequest.category,
            ),
        )

        return ItemResponse(item)
    }

    fun findItems(pageable: Pageable, name: String?) = itemRepository
        .findAllWithJpql(pageable, name)
        .map { ItemResponse(it) }

    fun findItem(id: Long): ItemResponse {
        val item = itemRepository.findByIdOrNull(id)
            ?: throw DomainException(ResponseMessage.DATA_NOT_FOUND, HttpStatus.NOT_FOUND)

        return ItemResponse(item)
    }

    @Transactional
    fun putItem(id: Long, itemSaveRequest: ItemSaveRequest): ItemResponse {
        val item = itemRepository.findByIdOrNull(id)?.apply {
            update(
                itemSaveRequest.name,
                itemSaveRequest.description,
                itemSaveRequest.price,
                itemSaveRequest.quantity,
                itemSaveRequest.category,
            )
        } ?: itemRepository.save(
            Item.create(
                itemSaveRequest.name,
                itemSaveRequest.description,
                itemSaveRequest.price,
                itemSaveRequest.quantity,
                itemSaveRequest.category,
            ),
        )

        return ItemResponse(item)
    }

    @Transactional
    fun deleteItem(id: Long) = itemRepository.deleteById(id)
}
